pkg/backporter/httpcache.go

```diff
@@ -0,0 +1,146 @@
+}
+
+func (bc *bugzillaCache) set(key string, respBytes []byte) {
+	bc.lock.Lock()
+	defer bc.lock.Unlock()
+	bc.cache[key] = respBytes
+}
+
+func newBugzillaCache() *bugzillaCache {
+	return &bugzillaCache{cache: map[string][]byte{}}
+}
+
+// cachingTransport is an implementation http.RoundTripper
+// which first checks for cached values
+type cachingTransport struct {
+	cache     *bugzillaCache
+	transport http.RoundTripper
+}
+
+// RoundTrip will first check if there are any cached responses and return that
+// if not it will make an HTTP call using the DefaultTransport
+func (t *cachingTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	if req.Method != "GET" {
+		return t.transport.RoundTrip(req)
+	}
+	var err error
+	var resp *http.Response
+	g := new(errgroup.Group)
+	g.Go(func() error {
+		resp, err = t.transport.RoundTrip(req)
+		if err != nil {
+			return err
+		}
+		if resp.StatusCode == http.StatusOK || resp.StatusCode == http.StatusNotFound {
+			body, err := httputil.DumpResponse(resp, true)
+			if err != nil {
+				return fmt.Errorf("err while serializing response to cache: %w", err)
+			}
+			t.cache.set(req.URL.String(), body)
+		}
+
+		return nil
+	})
+
+	if cachedVal, isCached := t.cache.get(req.URL.String()); isCached {
+		b := bytes.NewBuffer(cachedVal)
+		cachedResp, err := http.ReadResponse(bufio.NewReader(b), req)
+		if err != nil {
+			return nil, err
+		}
+		cachedResp.Header.Set(xFromCache, "1")
+		return cachedResp, nil
+	}
+	if err := g.Wait(); err != nil {
+		return nil, err
+	}
+	return resp, nil
+}
+
+func refreshCache(bc *bugzillaCache, m prometheus.Histogram) {
+	concurrencyLimiter := make(chan string, 10)
+	logrus.Infof("refreshing %d entries in cache", len(bc.cache))
+	m.Observe(float64(len(bc.cache)))
+	for url := range bc.cache {
+		concurrencyLimiter <- url
+		go func() {
+			url := <-concurrencyLimiter
+			resp, err := http.Get(url)
+			if err != nil {
+				logrus.WithError(fmt.Errorf("cache refresh error - failed to fetch %s: %w", url, err))
+				return
+			}
+			defer resp.Body.Close()
+			if resp.StatusCode == http.StatusOK || resp.StatusCode == http.StatusNotFound {
+				body, err := httputil.DumpResponse(resp, true)
+				if err != nil {
+					logrus.WithError(fmt.Errorf("cache refresh error - DumpResponse failed %s: %w", url, err))
+					return
+				}
+				bc.set(url, body)
+			}
+		}()
+	}
+
+}
+
+// NewCachingTransport is a constructor for cachingTransport
+// If an entry is present in the cache, it is immediately returned
+// while also generating an async HTTP call to the bugzilla server to get the latest value
+// which is stored in the cache.
+// Therefore this cache does *NOT* reduce the HTTP traffic, and is only used to speed up the response.
+func NewCachingTransport() http.RoundTripper {
+	t := cachingTransport{
+		cache:     newBugzillaCache(),
+		transport: http.DefaultTransport,
+	}
+	cacheRefreshMetrics := prometheus.NewHistogram(
+		prometheus.HistogramOpts{
```

Comment:
This is not a histogram, its a `counter`: https://prometheus.io/docs/concepts/metric_types/

Reply:
I do not understand why, could you please explain that?
Since I was visualizing that the metrics would keep track of how many bugs were refreshed during each refresh and not just being incremented everytime a call was made.

Reply:
Check out the link above, it has some nice explanations for the different types